Refactor CRD generation in Makefile

christarazi opened the pull request:
- Clarify function for all Cilium CRDs
- Makefile: Extract paths for CRD into variable
- Makefile: Extract CRD groups from generate-k8s-api
- Makefile: Refactor CRD generation to avoid copy-paste pattern
- Makefile: Clean up old CRD state, ensure blank slate

christarazi committed:
Makefile: Extract paths for CRD into variable

This cleans up the invocation of controller-gen so that any new paths
don't have to directly change the command line.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
Makefile, docs: Refactor CRD generation to avoid copy-paste pattern

Now, all a developer is expected to do is to add the name of the CRD
they want to generate into the corresponding variable (v2, v2alpha1,
etc), and the automation will handle it from there.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
Makefile: Extract CRD groups from generate-k8s-api

This will allow the content of the generate-k8s-api to stay the same
when needing to add another CRD group.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
Clarify function for all Cilium CRDs

This clarifies that this function will enumerate all Cilium CRDs,
whereas previously it was unclear which CRDs because it said "all CRDs".

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on Makefile:
```diff
@@ -245,24 +245,28 @@ GIT_VERSION: force
+CRDS_CILIUM_V2ALPHA1 ?= ciliumendpointslices \
+                        ciliumbgppeeringpolicies \
+                        ciliumloadbalancerippools \
+                        ciliumnodeconfigs
```
Use `:=` instead of `?=`

joestringer reviewed (approved):
Broadly looks good to me, how do we test that this doesn't break CRD generation for some of the files? (I'm sure you've done a fine job, but Makefile syntax can be finnicky and surprising at the best of times)

mhofstetter commented on Documentation/contributing/development/introducing_new_crds.rst:
We have to update the section about "Generating CRD YAML" too. (https://docs.cilium.io/en/v1.13/contributing/development/introducing_new_crds/#generating-crd-yaml)

Especially the point of having to update the `Makefile` itself.

> Update the ``Makefile`` to copy your generated CRD from a ``tmp`` directory to 
  the correct location in Cilium repository. Edit the following location

mhofstetter commented on Makefile:
```diff
@@ -245,24 +245,28 @@ GIT_VERSION: force
-	mv ${TMPDIR}/cilium.io_ciliumbgppeeringpolicies.yaml ./pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumbgppeeringpolicies.yaml
-	mv ${TMPDIR}/cilium.io_ciliumloadbalancerippools.yaml ./pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumloadbalancerippools.yaml
-	mv ${TMPDIR}/cilium.io_ciliumnodeconfigs.yaml ./pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml
+	for file in $(CRDS_CILIUM_V2); do mv ${TMPDIR}/cilium.io_$${file}.yaml ./pkg/k8s/apis/cilium.io/client/crds/v2/$${file}.yaml; done
```
we could get rid of the crd-to-version mapping with the following command which extracts the necessary information from the YAML itself. (`yq` required)

```
for file in ${TMPDIR}/*.yaml; do mv $${file} ./pkg/k8s/apis/cilium.io/client/crds/`yq '.spec.versions.0.name' $${file}`/`yq '.spec.names.plural' $${file}`.yaml; done
```

of course this only works as long as a CRD only has one version - which is kind of against the concept of a CRD. but we're "disrespecting" this anyway when trying to move a CRD into a specific version folder.

otherwise we could try to get rid of mapping it to a decicated version folder at all - but i kind of don't have the whole cilium history with CRDs :)

christarazi replied:
I thought about this (using `yq`) but didn't want to involve depending on a new dependency for this refactor. I also thought about an alternative to use some combination of more standard tools like `grep`, but thought that it wasn't really worth the complexity in the end. We don't add new CRDs that often. Additionally, I think the assumption that a developer already knows what API version their new CRD is going to be under is a safe one to make. Given that, it's just a matter of the developer updating the corresponding variable with their CRD name and then running the target. I didn't see how the additional complexity was beneficial enough for something like this. 

WDYT?

mhofstetter replied:
definitely agree on using as few external tools as possible in the Makefile - therefore perfectly fine with me!

christarazi committed:
Makefile: Clean up old CRD state, ensure blank slate on each run

With this, we can also detect stale generated artifacts. This is now
consistent with how the generate-k8s-api target works as well.

Suggested-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

mhofstetter commented on Makefile:
```diff
@@ -245,24 +245,34 @@ GIT_VERSION: force
-	mv ${TMPDIR}/cilium.io_ciliumnodeconfigs.yaml ./pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml
+
+	# Clean up old CRD state and start with a blank state.
+	for path in $(CRDS_CILIUM_PATHS); do rm -rf $${path} && mkdir $${path}; done
```
even though i meant to add it [here](https://github.com/cilium/cilium/blob/master/contrib/scripts/check-k8s-code-gen.sh#L6-L9) it's kind of nice to have it in the make target directly! helps to detect stale YAML (of potentially deleted / moved (version) CRDs) even sooner during development.  :+1:

qmonnet commented on Makefile:
```diff
@@ -265,6 +267,10 @@ manifests: ## Generate K8s manifests e.g. CRD, RBAC etc.
 	$(QUIET)$(GO) run sigs.k8s.io/controller-tools/cmd/controller-gen $(CRD_OPTIONS) paths=$(CRD_PATHS) output:crd:artifacts:config="$(TMPDIR)"
 	$(QUIET)$(GO) run ./tools/crdcheck "$(TMPDIR)"
+
+	# Clean up old CRD state and start with a blank state.
+	for path in $(CRDS_CILIUM_PATHS); do rm -rf $${path} && mkdir $${path}; done
```
Some nitpick suggestions but they don't really matter, I can't find a case where even `mkdir -p` would be necessary. Just “good practices” I guess.

```suggestion
	# Clean up old CRD state and start with a blank slate.
	for path in $(CRDS_CILIUM_PATHS); do $(RM) -rf -- $${path} && mkdir -p $${path}; done
```

qmonnet reviewed (approved):
Thanks!
